model/entry.go

```diff
@@ -0,0 +1,75 @@
+package model
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+type Entry struct {
+	Ts   time.Time
+	Line string
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (e *Entry) MarshalJSON() ([]byte, error) {
+	l, err := json.Marshal(e.Line)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(fmt.Sprintf("[\"%d\",%s]", e.Ts.UnixNano(), l)), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (e *Entry) UnmarshalJSON(data []byte) error {
+	var unmarshal []string
+
+	err := json.Unmarshal(data, &unmarshal)
+	if err != nil {
+		return err
+	}
+
+	t, err := strconv.ParseInt(unmarshal[0], 10, 64)
+	if err != nil {
+		return err
+	}
+
+	e.Ts = time.Unix(0, t)
+	e.Line = unmarshal[1]
+
+	return nil
+}
+
+type IntEntry struct {
+	Ts   int64
+	Line string
+}
+
+type Entries []Entry
+
+type LabeledEntry struct {
+	Ts     time.Time
```

Comment:
This should embed `Entry` instead of reusing the same fields, both for consumption (we don't have to reconstruct the `Entry` - we can just use `LabeledEntry.Entry`) and for maintainability.

I'd suggest
```go
type LabeledEntry struct {
    Entry
    Labels model.LabelSet
}
```